### PR TITLE
[ISSUE #10973] Reduce allocation in ExtraInfoUtil parsing and primitive getters

### LINKED ISSUE #10973: [Enhancement] Reduce allocation in ExtraInfoUtil POP info parsing and return primitives from numeric getters
### Before Creating the Enhancement Request

- [x] I have confirmed that this should be classified as an enhancement rather than a bug/feature.

### Summary

Reduce per-request allocation in `ExtraInfoUtil`: parse POP `startOffsetInfo` / `msgOffsetInfo` / `orderCountInfo` without the intermediate `split` arrays, and return primitives from the numeric extraInfo getters.

### Motivation

Every POP response and every ACK goes through `ExtraInfoUtil`:

1. `parseStartOffsetInfo` / `parseMsgOffsetInfo` / `parseOrderCountInfo` split each entry with `one.split(KEY_SEPARATOR)` (a `String[]` plus three substrings per entry) and re-concatenate the first two fields into a map key; `parseMsgOffsetInfo` additionally splits the offset list with `split(",")`.
2. `getCkQueueOffset` / `getPopTime` / `getInvisibleTime` return boxed `Long` although every caller in the repository immediately assigns the result to a `long` (broker ACK/changeInvisibleTime paths, client batch-ack path, proxy `LocalMessageService`).

### Solution

- Walk each entry with `indexOf` instead of `split`: no intermediate array, and the map key is built with `StringBuilder.append(CharSequence, int, int)` without substrings. The value field keeps one substring for `parseLong` (Java 8 target has no range-parse API).
- Change the three getters to return `long` (`Long.parseLong` instead of `Long.valueOf`). This is source-compatible for all in-repo callers (none rely on nullability or identity); it is binary-incompatible for externally compiled bytecode, which needs a recompile.
- Behavior note: the previous split-based validation half-accepted corrupt entries with empty fields or trailing separators (producing keys like `"@a"`); the new validation rejects them with the same `IllegalArgumentException` used for other malformed shapes. Well-formed wire strings produced by the builders are unaffected, covered by new round-trip tests.

### Verification

- `ExtraInfoUtilTest` extended with round-trip (normal + retry topic, multi-entry), getter, and malformed-input cases; 5/5 pass. `AckMessageProcessorTest` / `ChangeInvisibleTimeProcessorTest` / `PopMessageProcessorTest` 25/25, `MQClientAPIImplTest` 133/133.
- 4-node cluster A/B in POP mode (`mqadmin setConsumeMode -m POP`, producer 64 threads + consumer 20 threads, consume TPS steady at 150–154k, 3 interleaved trials per side):
  - broker side (remoting jar swapped on broker): young GC per million consumed msgs 2.60/2.62/2.70 (base) vs 2.63/2.64/2.62 (patch) — parity;
  - client side (remoting jar swapped on consumer): 1.10/1.08/1.10 vs 1.11/1.10/1.10 — parity.
  No regression on either side; the allocation saving itself (tens of bytes per response) is below GC-count resolution, so this is submitted as a cleanup-level optimization on a hot path.

### PR DESCRIPTION
**Pull request by wang-jiahua**
### Which Issue(s) This PR Fixes

Fixes #10973

### Brief Description

Every POP response and every ACK goes through `ExtraInfoUtil`. Two allocation sources removed without touching the wire format:

1. `parseStartOffsetInfo` / `parseMsgOffsetInfo` / `parseOrderCountInfo` used `split` per entry (a `String[]` plus three substrings), then re-concatenated the first two fields into a map key; `parseMsgOffsetInfo` split the offset list again with `split(",")`. Entries are now walked with `indexOf` and the map key is built with `StringBuilder.append(CharSequence, int, int)`, so the arrays and most substrings are gone (the value keeps one substring for `parseLong`; the Java 8 target has no range-parse API).
2. `getCkQueueOffset` / `getPopTime` / `getInvisibleTime` returned boxed `Long` although every caller in the repository assigns the result straight to a `long`. They now return `long`.

Notes for reviewers:

- The getter change is source-compatible for all in-repo callers (none rely on nullability or identity) but binary-incompatible for externally compiled bytecode, which needs a recompile against the new remoting artifact.
- The previous split-based validation half-accepted corrupt entries with empty fields or trailing separators (producing keys like `"@a"`); the new validation rejects them with the same `IllegalArgumentException` used for other malformed shapes. Builder-produced wire strings are unaffected, covered by new round-trip tests.

### How Did You Test This Change?

- `ExtraInfoUtilTest` extended with round-trip (normal + retry topic, multi-entry), getter, and malformed-input cases, 5/5 pass; `AckMessageProcessorTest` / `ChangeInvisibleTimeProcessorTest` / `PopMessageProcessorTest` 25/25; `MQClientAPIImplTest` 133/133.
- 4-node cluster A/B in POP mode (`mqadmin setConsumeMode -m POP`, producer 64 threads + consumer 20 threads, consume TPS steady at 150-154k, pop path confirmed active via pop.log, 3 interleaved trials per side):
  - broker side (remoting jar swapped on the broker): young GC per million consumed msgs 2.60/2.62/2.70 (base) vs 2.63/2.64/2.62 (patch) — parity;
  - client side (remoting jar swapped on the consumer): 1.10/1.08/1.10 vs 1.11/1.10/1.10 — parity.

No regression on either side; the allocation saving itself (tens of bytes per response) is below GC-count resolution, so this is a cleanup-level optimization on a hot path plus stricter rejection of corrupt extraInfo input.

### CODE REVIEW

**Review by RockteMQ-AI — APPROVED** [review]
﻿### Summary
This PR reduces allocations in `ExtraInfoUtil` by switching getter return types from boxed `Long` to primitive `long` (using `Long.parseLong` instead of `Long.valueOf`), and replacing `String.split()` with manual `indexOf`-based parsing in `parseMsgOffsetInfo`. The new `locateEntrySeparators` method packs two separator positions into a single `long` to avoid allocating intermediate `String[]` arrays, and `buildEntryKey` constructs the map key with a pre-sized `StringBuilder`.

The changes are correct and well-structured. The `locateEntrySeparators` validation logic properly handles edge cases (empty fields, too many separators). The return type change from `Long` to `long` is safe since these methods throw on invalid input and never returned null. Tests cover the round-trip behavior and malformed input scenarios.

LGTM — solid allocation reduction on a hot parsing path.

---
*Automated review by github-manager-bot*

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/header/ExtraInfoUtil.java [hunk 1]
```diff
@@ -216,18 +216,27 @@ public static Map<String, List<Long>> parseMsgOffsetInfo(String msgOffsetInfo) {
         }
 
         for (String one : array) {
-            String[] split = one.split(MessageConst.KEY_SEPARATOR);
-            if (split.length != 3) {
+            long separators = locateEntrySeparators(one);
+            if (separators < 0) {
                 throw new IllegalArgumentException("parse msgOffsetMap error, " + msgOffsetMap);
```
[the remark this answers is not in the document]
Comment (wang-jiahua) [reply remoting/src/main/java/org/apache/rocketmq/remoting/protocol/header/ExtraInfoUtil.java]: Agreed — both throw sites now include the original `msgOffsetInfo` string instead of the partially-built map (commit da87233).

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/header/ExtraInfoUtil.java [hunk 2]
```diff
@@ -274,20 +285,46 @@ public static Map<String, Integer> parseOrderCountInfo(String orderCountInfo) {
...
 
         return startOffsetMap;
     }
 
+    /**
+     * Locates the two {@link MessageConst#KEY_SEPARATOR} positions of an entry laid out as
+     * {@code retryFlag queueId value}, packed as {@code (sep1 << 32) | sep2}. Returns a negative
+     * value when the entry does not have exactly three non-empty fields, mirroring the previous
+     * split-based validation without allocating the intermediate array.
+     */
```
[the remark this answers is not in the document]
Comment (wang-jiahua) [reply remoting/src/main/java/org/apache/rocketmq/remoting/protocol/header/ExtraInfoUtil.java:311]: Reworded the Javadoc to state explicitly that this is stricter than the split-based validation and that corrupt entries are rejected with the same IllegalArgumentException the callers already throw (commit da87233).

**Review by RockteMQ-AI — APPROVED** [review]
### Summary
Reduces allocation in `ExtraInfoUtil` by (1) changing return types from `Long` to `long` (avoiding autoboxing), (2) replacing `split()` with manual `indexOf` parsing, and (3) using `Long.parseLong()` instead of `Long.valueOf()`.

The new `locateEntrySeparators()` is well-designed — stricter than the old `split` approach (rejects empty fields/trailing separators), and packing two positions into a `long` avoids a return-object allocation. The `buildEntryKey()` with capacity-hinted `StringBuilder` is a nice touch.

Good test updates to cover the stricter validation.

LGTM.

---
*Automated review by github-manager-bot*
